---
layout: page
title: Reproducible research tools course BIOS 692
tagline: Reproducible research tools course BIOS 692
description: Reproducible research tools course BIOS 692
---

Web pages for the BIOS 692 summer course offered by [Mikhail Dozmorov](https://mdozmorov.github.io/) in [the Department of Biostatistics](http://www.biostatistics.vcu.edu/), VCU.

BIOS 692-802 (1 cr.) (36610)

June 13, 14, 15 and 16 Monday – Thursday 9:00 – 10:50 a.m. OCS, rm. 5009

[Syllabus](pages/syllabus.html)

This course on GitHub [https://github.com/mdozmorov/BIOS692](https://github.com/mdozmorov/BIOS692)

I thank [Karl Broman](http://kbroman.org/), [Greg Wilson](http://sensibleadventures.com/) and [Software Carpentry](http://software-carpentry.org/), folks from [RStudio](https://www.rstudio.com/about/), and many others for providing material, tools and inspiration for creating this course

# Resources

## General

["Tools for Reproducible Research" course by Karl Broman](http://kbroman.org/Tools4RR/)

[Resources, "Steps towards reproducible research" ](http://kbroman.org/steps2rr/pages/resources.html)

[Software Carpentry' lessons](http://software-carpentry.org/lessons/)

[Software Carpentry' reading material on software engineering and scientific computing](http://software-carpentry.org/reading/)

[UW-Madison Software Carpentry Workshop for the Social Sciences](https://github.com/UW-Madison-ACI/boot-camps)

[Linux basics](https://sites.google.com/a/bioinformatics.ucr.edu/bioinformatics-manuals/home/linux-basics) and [R basics](https://sites.google.com/a/bioinformatics.ucr.edu/bioinformatics-manuals/home/R_BioCondManual) manuals by [Thomas Girke](http://girke.bioinformatics.ucr.edu/)

## Steps in reproducible research

### Overview

["Statistics for biologists" one-pagers about methods and reproducibility](http://www.nature.com/collections/qghhqm)

[Sandve, Geir Kjetil, Anton Nekrutenko, James Taylor, and Eivind Hovig. “Ten Simple Rules for Reproducible Computational Research.” PLoS Computational Biology 9, no. 10 (October 2013): e1003285. doi:10.1371/journal.pcbi.1003285.](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285)

### Linux/bash basics

[Survival guide for Unix newbies](http://matt.might.net/articles/basic-unix/)

[Settling into Unix](http://matt.might.net/articles/settling-into-unix/)

[Shell programming with bash](http://matt.might.net/articles/bash-by-example/)

[CommandLineFu.com, command-line one-liner gems](http://www.commandlinefu.com/commands/browse)

["The Unix shell", Software Carpentry](https://swcarpentry.github.io/shell-novice/)

### Text manipulation with grep, awk, sed, vim

[Tutorial to sed by Bruce Barnett](http://www.grymoire.com/Unix/Sed.html)

## Automating everything

### Best practices of data/code organization

[Organize your project](http://kbroman.org/steps2rr/pages/organize.html)

[Anti-patterns, Clean Code and Beyond](http://www.cbs.dtu.dk/courses/27610/clean_code_index.html)

[Ten Simple Rules for Robustifying Your Software](https://github.com/oicr-gsi/robust-paper)

["Code and Data for the Social Sciences: A Practitioner’s Guide" book by Matthew Gentzkow and Jesse Shapiro, PDF](https://web.stanford.edu/~gentzkow/research/CodeAndData.pdf)

[Wilson, Greg, D. A. Aruliah, C. Titus Brown, Neil P. Chue Hong, Matt Davis, Richard T. Guy, Steven H. D. Haddock, et al. “Best Practices for Scientific Computing.” PLoS Biology 12, no. 1 (January 2014): e1001745. doi:10.1371/journal.pbio.1001745.](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745)

### Makefiles

["A minimal tutorial on make" by Karl](http://kbroman.org/minimal_make/)

["Learning about Makefiles" by Dave Tang](http://davetang.org/muse/2015/05/31/learning-about-makefiles/)

### RStudio, R functions & packages

[Repeated code into functions](http://kbroman.org/steps2rr/pages/functions.html)

[Package functions](http://kbroman.org/steps2rr/pages/packages.html)

[Package tutorial by Hillary Parker](https://hilaryparker.com/2014/04/29/writing-an-r-package-from-scratch/)

[R package primer by Karl Broman](http://kbroman.org/pkg_primer/)

["R packages" book by Hadley Wickham](http://r-pkgs.had.co.nz/)

[Jeff Leek on developing R packages](https://github.com/jtleek/rpackages)

## Reproducible reports

### Literate programming with Markdown/KnitR

["Turn scripts into reproducible reports" by Karl Broman](http://kbroman.org/steps2rr/pages/reports.html)

["Markdown" by Karl Broman](http://kbroman.org/knitr_knutshell/pages/markdown.html)

["R Markdown" by Karl](http://kbroman.org/knitr_knutshell/pages/Rmarkdown.html)

["A quick introduction to R/markdown" presentation by Peter Ralph](http://petrelharp.github.io/r-markdown-tutorial/using-rmarkdown.slides.html), and [some R Markdown gotchas (advanced)](https://petrelharp.github.io/r-markdown-tutorial/gotchas.html)

[R Markdown guides from Rstudio](https://support.rstudio.com/hc/en-us/articles/205368677-R-Markdown-Dynamic-Documents-for-R)

### Data manipulation and visualization in R

[Data Manipulation Using R (& dplyr)](http://www.slideshare.net/Ram-N/data-manipulation-using-r-acm2014)

[Data Manipulation with dplyr](http://datascienceplus.com/data-manipulation-with-dplyr/)

["Aggregating and analyzing data with dplyr" by Data Carpentry](http://www.datacarpentry.org/R-genomics/04-dplyr.html)

[Do your "data janitor work" like a boss with dplyr](http://www.gettinggeneticsdone.com/2014/08/do-your-data-janitor-work-like-boss.html)

["Data visualization in R" by Data Carpentry](http://www.datacarpentry.org/R-genomics/05-data-visualization.html)

### Reproducible presentation and web-publishing

[Easy web publishing from R on Rpubs.com](http://rpubs.com/)

[Slidify: Modern, simple presentations written in R Markdown](https://benjaminlmoore.wordpress.com/2014/02/24/slidify-presentations-in-r-markdown/)

["Easy websites with GitHub Pages" by Karl](http://kbroman.org/simple_site/)

[Creating web documentation with Bookdown](https://github.com/rstudio/bookdown), [Example, "Bookdown: Authoring Books with R Markdown" by Yihui Xie](https://bookdown.org/yihui/bookdown/)

## Version control, sharing and collaboration

### Git/GitHub

[One-pager simple git guide](https://rogerdudler.github.io/git-guide/)

[One-pager of git commands](https://github.com/kbroman/Tools4RR/blob/master/04_Git/GitCommands/git_notes.md)

[Interactive git](https://try.github.io/levels/1/challenges/1)

[Interactive git branching tutorial](http://learngitbranching.js.org/)

[Quick Git and GitHub videos](http://www.dataschool.io/git-and-github-videos-for-beginners/)

["Git and GitHub guide" by Karl Broman](http://kbroman.org/github_tutorial/)

[Software Carpentry course](https://swcarpentry.github.io/git-novice/)

[GitHub training videos](https://www.youtube.com/user/GitHubGuides/videos)

[Book "Version Control by Example" by Eric Sink](http://ericsink.com/vcbe/)

### Licenses

["Licensing", SoftwareCarpentry](https://swcarpentry.github.io/git-novice/11-licensing.html)

["Pick a License, Any License"](https://blog.codinghorror.com/pick-a-license-any-license/)

["License your software" by Karl Broman](http://kbroman.org/steps2rr/pages/licenses.html)

[Morin, Andrew, Jennifer Urban, and Piotr Sliz. “A Quick Guide to Software Licensing for the Scientist-Programmer.” PLoS Computational Biology 8, no. 7 (2012): e1002598. doi:10.1371/journal.pcbi.1002598.](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002598)

### Data/code sharing repositories

[Goodman, Alyssa, Alberto Pepe, Alexander W. Blocker, Christine L. Borgman, Kyle Cranmer, Merce Crosas, Rosanne Di Stefano, et al. “Ten Simple Rules for the Care and Feeding of Scientific Data.” PLoS Computational Biology 10, no. 4 (April 2014): e1003542. doi:10.1371/journal.pcbi.1003542.](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003542) - lists all major data sharing repositories
